api/src/api.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "api/inc/uvisor-lib.h"
+#include <stdint.h>
+
+
+/******************************** namespace ********************************/
+
+int uvisor_box_namespace(int box_id, char *box_namespace, size_t length)
+{
+    return uvisor_api.box_namespace(box_id, box_namespace, length);
+}
+
+/********************************** debug **********************************/
+void uvisor_debug_init(const TUvisorDebugDriver * const driver)
+{
+    uvisor_api.debug_init(driver);
+}
+
+void uvisor_error(THaltUserError reason)
+{
+    uvisor_api.error(reason);
+}
+
+/******************************** page heap ********************************/
+int uvisor_page_malloc(UvisorPageTable * const table)
+{
+    return uvisor_api.page.malloc(table);
```

Comment:
A benefit of uVisor core doing the transition is that interface between uVisor core and all other software is just a simple table of functions that are callable via the ordinary AAPCS. uVisor core is then able to later decide how the transitions should work and when, which is of benefit when we are still trying out different models of transitioning.

Also note that not all entry points into uVisor core will perform a transition: pool queue code is shared between uVisor core and other software, but lives in the uVisor core binary because uVisor core also runs the code and we want all code that runs privileged to live in the uVisor core binary.

2 sounds like a "nice to have" and not a "must have", but I don't see the basic approach we are taking here as antithetical to 2; we could still do 2, but where an API runs is platform dependent (v7 vs v8) and awkward to annotate at this level of abstraction.